### PR TITLE
NewCrop: support for 4D inputs

### PR DESCRIPTION
**Pull request by jantonguirao**

### CODE REVIEW

File: dali/pipeline/operators/crop/new_crop.cc [hunk 1]
```diff
@@ -34,21 +35,32 @@ DALI_SCHEMA(NewCrop)
...
+  VALUE_SWITCH(number_of_dims, NumDims, (3, 4), (
+    kernels::KernelContext ctx;
+    auto in_view = view<const InputType, NumDims>(input);
+
+    std::array<int64_t, NumDims> anchor, shape;
+    for (std::size_t d = 0; d < NumDims; d++) {
+      anchor[d] = slice_anchor[d];
+      shape[d] = slice_shape[d];
+    }
+    kernels::SliceArgs<NumDims> slice_args = {std::move(anchor), std::move(shape)};
```
Comment (klecki) [inline dali/pipeline/operators/crop/new_crop.cc]: What are those moves supposed to do? 

You could create empty SliceArgs, and fill it directly as in loop above.

File: dali/pipeline/operators/crop/new_crop.cc [hunk 2]
```diff
@@ -34,21 +35,32 @@ DALI_SCHEMA(NewCrop)
...
+
+    kernels::SliceCPU<OutputType, InputType, NumDims> kernel;
+    kernels::KernelRequirements kernel_req = kernel.Setup(ctx, in_view, slice_args);
```
Comment (klecki) [inline dali/pipeline/operators/crop/new_crop.cc]: This is unused, and the output sizes are calculated independently.

File: dali/pipeline/operators/crop/new_crop.cc [hunk 3]
```diff
@@ -67,19 +79,22 @@ void NewCrop<CPUBackend>::DataDependentSetup(SampleWorkspace *ws, const int idx)
...
   auto data_idx = ws->data_idx();
   DataDependentSetup(data_idx, in_layout, input.shape());
   auto &slice_shape = slice_shapes_[data_idx];
-  Dims output_shape = { slice_shape[0], slice_shape[1], slice_shape[2] };
 
   auto &output = ws->Output<CPUBackend>(idx);
   output.SetLayout(out_layout);
-  output.Resize(output_shape);
+  if (in_layout == DALI_NFHWC || in_layout == DALI_NFCHW) {
+    output.Resize({slice_shape[0], slice_shape[1], slice_shape[2], slice_shape[3]});
```
Comment (klecki) [inline dali/pipeline/operators/crop/new_crop.cc]: Shouldn't it use the kernel's `Setup`?

File: dali/pipeline/operators/crop/new_crop.cu [hunk 4]
```diff
@@ -63,15 +77,20 @@ void NewCrop<GPUBackend>::DataDependentSetup(DeviceWorkspace *ws, const int idx)
...
     "Unexpected data layout");
   DALITensorLayout out_layout = in_layout;
 
   std::vector<Dims> output_shape(batch_size_);
   for (int i = 0; i < batch_size_; ++i) {
     DataDependentSetup(i, in_layout, input.tensor_shape(i));
     auto &slice_shape = slice_shapes_[i];
-    output_shape[i] = { slice_shape[0], slice_shape[1], slice_shape[2] };
+    if (in_layout == DALI_NFHWC || in_layout == DALI_NFCHW) {
+      output_shape[i] = { slice_shape[0], slice_shape[1], slice_shape[2], slice_shape[3] };
```
Comment (klecki) [inline dali/pipeline/operators/crop/new_crop.cu]: Same here

File: dali/pipeline/operators/crop/new_crop.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/new_crop.cc]: ok, I'll change that

File: dali/pipeline/operators/crop/new_crop.cc (hunk 2 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/new_crop.cc]: done

File: dali/pipeline/operators/crop/new_crop.cc (hunk 3 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/new_crop.cc]: done

File: dali/pipeline/operators/crop/new_crop.cu (hunk 4 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/new_crop.cu]: Done

**Comment by jantonguirao** [comment]
!build

**Pull request merged**